Compartmentalize how ProductRegistry is filled

Dr15Jones opened the pull request:
#### PR description:

- constrain where in the code the ProductRegistry has additional non-dropped entries added
- isolate when in the construction sequence a signal is sent when the ProductRegistry is changed

#### PR validation:

This is a code refactoring and as such no changes in behavior are expected. Code compiles and all but one framework unit test pass.

The only known failure is `TestIOPoolInputNoParentDictionary`. When that test run it copies parts of code from CMSSW to a new build area. The problem is it doesn't copy enough code to work with this change (code in DataFormats/Common/test fails to build).

resolves https://github.com/cms-sw/framework-team/issues/1296

Dr15Jones committed:
Hold ProductResolverIndexHelper via const

To set the helper, a temporary is created and filled and then moved to the ProductRegistry.

Dr15Jones committed:
Only SignallingProductRegistry can add products

This isolates where where the ProductRegistry can have additional products added.

Dr15Jones committed:
Must explicitly ask for the ProductRegistry in SignallingProductRegistry

Dr15Jones committed:
Move to SignallingProductRegistryFiller

Moved from SignallingProductRegistry. The new class now contains a ProductRegistry rather than inheriting from one. This allowed to addCalled method to be removed and fully implemented in the new class.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"692153cfe658e8d690856deff0241baabccd8bed":{"files":["DataFormats/Provenance/interface/ProductRegistry.h","DataFormats/Provenance/src/ProductRegistry.cc","FWCore/Framework/interface/SignallingProductRegistry.h","FWCore/Framework/src/EventProcessor.cc","FWCore/Framework/src/Schedule.cc","FWCore/Framework/src/ScheduleItems.cc","FWCore/Framework/src/StreamSchedule.cc","FWCore/Framework/src/SubProcess.cc","FWCore/Framework/src/processEDAliases.cc","FWCore/Framework/test/Event_t.cpp","FWCore/Framework/test/edproducer_productregistry_callback.cc","FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc","FWCore/Framework/test/eventprincipal_t.cppunit.cc","FWCore/Framework/test/generichandle_t.cppunit.cc","FWCore/Framework/test/productregistry.cppunit.cc","FWCore/Framework/test/test_catch2notTP_MergeableRunProductMetadata.cc","FWCore/Sources/src/PuttableSourceBase.cc"],"squashed":false,"time":1741356229},"a1bfffa95392ef1f4f70a3d9dd5946cfd873d0de":{"files":["DataFormats/Provenance/interface/ProductRegistry.h","DataFormats/Provenance/src/ProductRegistry.cc","FWCore/Framework/interface/EventProcessor.h","FWCore/Framework/interface/GlobalSchedule.h","FWCore/Framework/interface/ModuleChanger.h","FWCore/Framework/interface/OutputModuleCore.h","FWCore/Framework/interface/ProducerBase.h","FWCore/Framework/interface/ProductRegistryHelper.h","FWCore/Framework/interface/Schedule.h","FWCore/Framework/interface/ScheduleItems.h","FWCore/Framework/interface/SignallingProductRegistryFiller.h","FWCore/Framework/interface/StreamSchedule.h","FWCore/Framework/interface/WorkerManager.h","FWCore/Framework/interface/global/EDAnalyzerBase.h","FWCore/Framework/interface/global/EDFilterBase.h","FWCore/Framework/interface/global/EDProducerBase.h","FWCore/Framework/interface/limited/EDAnalyzerBase.h","FWCore/Framework/interface/limited/EDFilterBase.h","FWCore/Framework/interface/limited/EDProducerBase.h","FWCore/Framework/interface/maker/InputSourceFactory.h","FWCore/Framework/interface/maker/MakeModuleParams.h","FWCore/Framework/interface/maker/ModuleHolder.h","FWCore/Framework/interface/maker/WorkerParams.h","FWCore/Framework/interface/one/EDAnalyzerBase.h","FWCore/Framework/interface/one/EDFilterBase.h","FWCore/Framework/interface/one/EDProducerBase.h","FWCore/Framework/interface/stream/EDAnalyzerAdaptorBase.h","FWCore/Framework/interface/stream/EDAnalyzerBase.h","FWCore/Framework/interface/stream/ProducingModuleAdaptorBase.h","FWCore/Framework/src/EventProcessor.cc","FWCore/Framework/src/GlobalSchedule.cc","FWCore/Framework/src/ModuleChanger.cc","FWCore/Framework/src/OutputModuleCore.cc","FWCore/Framework/src/ProducerBase.cc","FWCore/Framework/src/ProductRegistryHelper.cc","FWCore/Framework/src/Schedule.cc","FWCore/Framework/src/ScheduleItems.cc","FWCore/Framework/src/SignallingProductRegistryFiller.cc","FWCore/Framework/src/StreamSchedule.cc","FWCore/Framework/src/WorkerMaker.cc","FWCore/Framework/src/WorkerManager.cc","FWCore/Framework/src/global/EDAnalyzerBase.cc","FWCore/Framework/src/limited/EDAnalyzerBase.cc","FWCore/Framework/src/one/EDAnalyzerBase.cc","FWCore/Framework/src/processEDAliases.cc","FWCore/Framework/src/processEDAliases.h","FWCore/Framework/src/stream/EDAnalyzerAdaptorBase.cc","FWCore/Framework/src/stream/EDAnalyzerBase.cc","FWCore/Framework/src/stream/ProducingModuleAdaptorBase.cc","FWCore/Framework/test/Event_t.cpp","FWCore/Framework/test/edproducer_productregistry_callback.cc","FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc","FWCore/Framework/test/eventprincipal_t.cppunit.cc","FWCore/Framework/test/generichandle_t.cppunit.cc","FWCore/Framework/test/maker2_t.cppunit.cc","FWCore/Framework/test/productregistry.cppunit.cc","FWCore/Framework/test/test_catch2_Factory.cc","FWCore/Framework/test/test_catch2notTP_MergeableRunProductMetadata.cc","FWCore/Sources/src/PuttableSourceBase.cc","FWCore/TestProcessor/interface/TestProcessor.h","FWCore/TestProcessor/interface/TestSourceProcessor.h","IOPool/SecondaryInput/test/SecondaryProducer.cc"],"squashed":false,"time":1741371800},"e3e58f911517a1eb18fad3a74567a2d3b43190d1":{"files":["DataFormats/Provenance/interface/ProductRegistry.h","DataFormats/Provenance/src/ProductRegistry.cc","FWCore/Framework/interface/EventProcessor.h","FWCore/Framework/interface/InputSource.h","FWCore/Framework/interface/ProductRegistryHelper.h","FWCore/Framework/interface/Schedule.h","FWCore/Framework/interface/SignallingProductRegistry.h","FWCore/Framework/interface/StreamSchedule.h","FWCore/Framework/src/EventProcessor.cc","FWCore/Framework/src/ProducerBase.cc","FWCore/Framework/src/ProductRegistryHelper.cc","FWCore/Framework/src/Schedule.cc","FWCore/Framework/src/StreamSchedule.cc","FWCore/Framework/src/SubProcess.cc","FWCore/Framework/src/processEDAliases.cc","FWCore/Framework/src/processEDAliases.h","FWCore/Framework/test/Event_t.cpp","FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc","FWCore/Framework/test/eventprincipal_t.cppunit.cc","FWCore/Framework/test/generichandle_t.cppunit.cc","FWCore/Framework/test/productregistry.cppunit.cc","FWCore/Framework/test/test_catch2notTP_MergeableRunProductMetadata.cc","FWCore/TestProcessor/src/TestProcessor.cc","FWCore/TestProcessor/src/TestSourceProcessor.cc","IOPool/SecondaryInput/test/SecondaryProducer.cc"],"squashed":false,"time":1741353829},"e7f919b25d3790e5398f4a33713fdda173a0e0b6":{"files":["DataFormats/Provenance/interface/ProductRegistry.h","DataFormats/Provenance/src/ProductRegistry.cc"],"squashed":false,"time":1741277625}},"emoji":{},"last_seen_sha":"a1bfffa95392ef1f4f70a3d9dd5946cfd873d0de","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43991

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Provenance/interface/ProductRegistry.h  modified in PR(s): #46858
  - File  DataFormats/Provenance/src/ProductRegistry.cc  modified in PR(s): #46858
  - File  FWCore/Framework/interface/GlobalSchedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/interface/ProductRegistryHelper.h  modified in PR(s): #47339
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/GlobalSchedule.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/ProductRegistryHelper.cc  modified in PR(s): #47397
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46858, #47467
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47467
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #46858
  - File  FWCore/TestProcessor/interface/TestProcessor.h  modified in PR(s): #47339
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47467

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43991/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43991/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43992

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Provenance/interface/ProductRegistry.h  modified in PR(s): #46858
  - File  DataFormats/Provenance/src/ProductRegistry.cc  modified in PR(s): #46858
  - File  FWCore/Framework/interface/GlobalSchedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/interface/ProductRegistryHelper.h  modified in PR(s): #47339
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/GlobalSchedule.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/ProductRegistryHelper.cc  modified in PR(s): #47397
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46858, #47467
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47467
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #46858
  - File  FWCore/TestProcessor/interface/TestProcessor.h  modified in PR(s): #47339
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47467

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones for master.

It involves the following packages:

- DataFormats/Provenance (**core**)
- FWCore/Framework (**core**)
- FWCore/Sources (**core**)
- FWCore/TestProcessor (**core**)
- IOPool/SecondaryInput (**core**)


@Dr15Jones, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @mmusich, @rovere, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**:   Build  HeaderConsistency  ClangBuild 
**Size**: This PR adds an extra 480KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44853/summary.html
**COMMIT**: 9eaf910cd2b43109506f0052ac7d4bf96fba7913
**CMSSW**: CMSSW_15_1_X_2025-03-07-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47533/44853/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-03-07-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-03-07-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_X_2025-03-07-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-94d5b66203ab8a015ad76ed29be3e51e/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/boost/1.80.0-7642db88b69862429c1c9c9980662789/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/curl/7.79.0-d8c0d5017cfad573d276c56bd08114b8/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/lcg/root/6.32.11-575650675fc09042e42e48da8061099a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tbb/v2022.0.0-10fdc07bd3d228f4c274203a8cd20b5c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/Mixing/Base/src/MixingBase/SecondaryEventProvider.cc.d src/Mixing/Base/src/SecondaryEventProvider.cc -o tmp/el8_amd64_gcc12/src/Mixing/Base/src/MixingBase/SecondaryEventProvider.cc.o
>> Compiling  src/Mixing/Base/src/PileupRandomNumberGenerator.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-03-07-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-03-07-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_X_2025-03-07-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/coral/CORAL_2_3_21-94d5b66203ab8a015ad76ed29be3e51e/include/LCG -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/boost/1.80.0-7642db88b69862429c1c9c9980662789/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/curl/7.79.0-d8c0d5017cfad573d276c56bd08114b8/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/lcg/root/6.32.11-575650675fc09042e42e48da8061099a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tbb/v2022.0.0-10fdc07bd3d228f4c274203a8cd20b5c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/Mixing/Base/src/MixingBase/PileupRandomNumberGenerator.cc.d src/Mixing/Base/src/PileupRandomNumberGenerator.cc -o tmp/el8_amd64_gcc12/src/Mixing/Base/src/MixingBase/PileupRandomNumberGenerator.cc.o
In file included from src/Mixing/Base/interface/PileUp.h:14,
                 from src/Mixing/Base/src/PileUp.cc:1:
poison/FWCore/Framework/interface/SignallingProductRegistry.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
    1 | #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
      |  ^~~~~
In file included from src/Mixing/Base/src/PileUp.cc:8:
poison/FWCore/Framework/interface/SignallingProductRegistry.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
    1 | #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43993

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Provenance/interface/ProductRegistry.h  modified in PR(s): #46858
  - File  DataFormats/Provenance/src/ProductRegistry.cc  modified in PR(s): #46858
  - File  FWCore/Framework/interface/GlobalSchedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/interface/ProductRegistryHelper.h  modified in PR(s): #47339
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/GlobalSchedule.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/ProductRegistryHelper.cc  modified in PR(s): #47397
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46858, #47467
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47467
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #46858
  - File  FWCore/TestProcessor/interface/TestProcessor.h  modified in PR(s): #47339
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47467

cmsbuild commented:
Pull request #47533 was updated. @Dr15Jones, @civanch, @cmsbuild, @kpedro88, @makortel, @mdhildreth, @smuzaffar can you please check and sign again.

Dr15Jones commented:
please test

makortel commented on DataFormats/Provenance/interface/ProductRegistry.h:
```diff
@@ -163,6 +143,30 @@ namespace edm {
+                                             std::string const& instanceAlias);
+
+    // triggers callbacks for modules watching registration
+    template <typename F>
+    void addFromInput_(edm::ProductRegistry const& iReg, F const& iCallback) {
```
If `const`-only access is not important, this could use the universal reference for `F`
```suggestion
    template <typename F>
    void addFromInput_(edm::ProductRegistry const& iReg, F&& iCallback) {
```
Universal reference would allow mutable functors.

Dr15Jones replied:
done

makortel commented on FWCore/Framework/interface/SignallingProductRegistryFiller.h:
```diff
@@ -0,0 +1,114 @@
+    void callForEachBranch(T const& iFunc) {
+      //NOTE: If implementation changes from a map, need to check that iterators are still valid
+      // after an insert with the new container, else need to copy the container and iterate over the copy
+      for (ProductRegistry::ProductList::const_iterator itEntry = registry_.productList().begin(),
+                                                        itEntryEnd = registry_.productList().end();
+           itEntry != itEntryEnd;
+           ++itEntry) {
+        if (itEntry->second.present()) {
+          iFunc(itEntry->second);
+        }
+      }
```
Could become
```suggestion
      for (auto const& entry : registry_.productList()) {
        if (entry.second.present()) {
          iFunc(entry.second);
        }
      }
```

Dr15Jones replied:
I was originally concerned that the _range based for loop_ might not have the same semantics as the iterator based one and I needed that semantics since the container might be updating while the loop is happening. A bit of research and I believe the semantics are retained so I changed the code.

makortel commented on FWCore/Framework/interface/SignallingProductRegistryFiller.h:
```diff
@@ -0,0 +1,114 @@
+    }
+    ProductRegistry::ProductList& productListUpdator() { return registry_.productListUpdator(); }
+
+    SignallingProductRegistryFiller(SignallingProductRegistryFiller const&) = delete;  // Disallow copying and moving
+    SignallingProductRegistryFiller& operator=(SignallingProductRegistryFiller const&) =
+        delete;  // Disallow copying and moving
```
I'd prefer to have the deletion of move operations to be explicit
```suggestion
    SignallingProductRegistryFiller(SignallingProductRegistryFiller const&) = delete;
    SignallingProductRegistryFiller& operator=(SignallingProductRegistryFiller const&) = delete;
    SignallingProductRegistryFiller(SignallingProductRegistryFiller&&) = delete;
    SignallingProductRegistryFiller& operator=(SignallingProductRegistryFiller&&) = delete;
```
and moved next to the constructors.

Dr15Jones replied:
done

makortel commented on FWCore/Framework/src/Schedule.cc:
```diff
@@ -605,6 +605,7 @@ namespace edm {
     std::map<std::string, std::vector<std::pair<std::string, int>>> outputModulePathPositions;
     reduceParameterSet(proc_pset, tns.getEndPaths(), modulesInConfig, usedModuleLabels, outputModulePathPositions);
     {
+      //CDJ: needs SignallingProductRegistryFiller
```
Just to double check, is this comment relevant to kept (as opposed to your note during development)?

Dr15Jones replied:
removed

makortel reviewed (commented):
Few things that caught my eye, **if** you feel like polishing further.

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 480KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44854/summary.html
**COMMIT**: 22cdb24bb8c7dfa4aba82dd50103d4f325594c6b
**CMSSW**: CMSSW_15_1_X_2025-03-07-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47533/44854/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin src/DQM/SiStripMonitorHardware/src/SiStripSpyEventMatcherModule.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-03-07-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-03-07-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_X_2025-03-07-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/dd4hep/v01-29-00-28723fa7011c94f5579247e3119e5ef7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/alpaka/1.2.0-0ef9ed9c17707f526f6e6d8263e2898d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/classlib/3.1.3-2827e9dbeb7dd2aced5a46b8a966dfab/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/boost/1.80.0-7642db88b69862429c1c9c9980662789/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/cuda/12.8.0-15bfa86985d46d842bb5ecc3aca6c676/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc/2.06.10-262d73524c32528e87adc31a98f68d52/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hls/2019.08-e6beae7d560007d8bb20c2cf88bfde9a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/protobuf/3.21.9-1126508a53768c90e66f6bf1821ac03a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/lcg/root/6.32.11-575650675fc09042e42e48da8061099a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tbb/v2022.0.0-10fdc07bd3d228f4c274203a8cd20b5c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/vdt/0.4.3-f008bd09702a81e8a062bbef698ed3af/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc3/3.2.7-f10dce5b16f0246f8cac8378933b9681/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyEventMatcherModule.cc.d src/DQM/SiStripMonitorHardware/src/SiStripSpyEventMatcherModule.cc -o tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyEventMatcherModule.cc.o
>> Compiling edm plugin src/DQM/SiStripMonitorHardware/src/SiStripSpyEventSummaryProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-03-07-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-03-07-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_X_2025-03-07-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/dd4hep/v01-29-00-28723fa7011c94f5579247e3119e5ef7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/alpaka/1.2.0-0ef9ed9c17707f526f6e6d8263e2898d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/classlib/3.1.3-2827e9dbeb7dd2aced5a46b8a966dfab/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/boost/1.80.0-7642db88b69862429c1c9c9980662789/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/cuda/12.8.0-15bfa86985d46d842bb5ecc3aca6c676/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc/2.06.10-262d73524c32528e87adc31a98f68d52/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hls/2019.08-e6beae7d560007d8bb20c2cf88bfde9a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/protobuf/3.21.9-1126508a53768c90e66f6bf1821ac03a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/lcg/root/6.32.11-575650675fc09042e42e48da8061099a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tbb/v2022.0.0-10fdc07bd3d228f4c274203a8cd20b5c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/vdt/0.4.3-f008bd09702a81e8a062bbef698ed3af/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc3/3.2.7-f10dce5b16f0246f8cac8378933b9681/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyEventSummaryProducer.cc.d src/DQM/SiStripMonitorHardware/src/SiStripSpyEventSummaryProducer.cc -o tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyEventSummaryProducer.cc.o
In file included from src/DQM/SiStripMonitorHardware/src/SiStripSpyEventMatcher.cc:9:
poison/FWCore/Framework/interface/SignallingProductRegistry.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
    1 | #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
      |  ^~~~~
>> Compiling edm plugin src/DQM/SiStripMonitorHardware/src/SiStripSpyExtractRunModule.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -DDD4HEP_USE_GEANT4_UNITS=1 -DCMSSW_GIT_HASH='CMSSW_15_1_X_2025-03-07-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_1_X_2025-03-07-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_X_2025-03-07-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/dd4hep/v01-29-00-28723fa7011c94f5579247e3119e5ef7/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/alpaka/1.2.0-0ef9ed9c17707f526f6e6d8263e2898d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/classlib/3.1.3-2827e9dbeb7dd2aced5a46b8a966dfab/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/boost/1.80.0-7642db88b69862429c1c9c9980662789/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/clhep/2.4.7.1-d3a3e353d370e701238f7949a0d7909f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/cuda/12.8.0-15bfa86985d46d842bb5ecc3aca6c676/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/gsl/2.6-f7574c606b0ce57ff601d3ca9534cd01/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc/2.06.10-262d73524c32528e87adc31a98f68d52/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hls/2019.08-e6beae7d560007d8bb20c2cf88bfde9a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/protobuf/3.21.9-1126508a53768c90e66f6bf1821ac03a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/lcg/root/6.32.11-575650675fc09042e42e48da8061099a/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tbb/v2022.0.0-10fdc07bd3d228f4c274203a8cd20b5c/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/cms/vdt/0.4.3-f008bd09702a81e8a062bbef698ed3af/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-5d91c922e771c0dc4f6bc00f61f3e2c5/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/hepmc3/3.2.7-f10dce5b16f0246f8cac8378933b9681/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/OpenBLAS/0.3.27-70a9dd2c9f309171934f13e3003b0540/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02879/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyExtractRunModule.cc.d src/DQM/SiStripMonitorHardware/src/SiStripSpyExtractRunModule.cc -o tmp/el8_amd64_gcc12/src/DQM/SiStripMonitorHardware/src/DQMSiStripMonitorHardware/SiStripSpyExtractRunModule.cc.o
>> Compiling edm plugin src/DQM/SiStripMonitorHardware/src/SiStripSpyIdentifyRuns.cc
</pre>

## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43995

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Provenance/interface/ProductRegistry.h  modified in PR(s): #46858
  - File  DataFormats/Provenance/src/ProductRegistry.cc  modified in PR(s): #46858
  - File  FWCore/Framework/interface/GlobalSchedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/interface/ProductRegistryHelper.h  modified in PR(s): #47339
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/GlobalSchedule.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/ProductRegistryHelper.cc  modified in PR(s): #47397
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46858, #47467
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47467
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #46858
  - File  FWCore/TestProcessor/interface/TestProcessor.h  modified in PR(s): #47339
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47467

cmsbuild commented:
Pull request #47533 was updated. @Dr15Jones, @antoniovagnerini, @civanch, @kpedro88, @makortel, @mdhildreth, @rseidita, @smuzaffar can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 60KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44859/summary.html
**COMMIT**: 6c3e6a3e587497892241aad81a9308f4392bb742
**CMSSW**: CMSSW_15_1_X_2025-03-08-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47533/44859/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47533/43998

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Provenance/interface/ProductRegistry.h  modified in PR(s): #46858
  - File  DataFormats/Provenance/src/ProductRegistry.cc  modified in PR(s): #46858
  - File  FWCore/Framework/interface/GlobalSchedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/interface/ProductRegistryHelper.h  modified in PR(s): #47339
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #47467
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/GlobalSchedule.cc  modified in PR(s): #47467
  - File  FWCore/Framework/src/ProductRegistryHelper.cc  modified in PR(s): #47397
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46858, #47467
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #47467
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #46858
  - File  FWCore/TestProcessor/interface/TestProcessor.h  modified in PR(s): #47339
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #47467

cmsbuild commented:
Pull request #47533 was updated. @Dr15Jones, @antoniovagnerini, @civanch, @kpedro88, @makortel, @mdhildreth, @rseidita, @smuzaffar can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 60KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44866/summary.html
**COMMIT**: 4c63b9624c5d00664adb5e488badce2aafcff1a6
**CMSSW**: CMSSW_15_1_X_2025-03-08-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47533/44866/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 2 errors in the following unit tests:

<pre>---> test TestIOPoolInputNoParentDictionary had ERRORS
---> test TestGeneratorInterfaceLHEInterfaceFileReading had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-08-1100+1b09ca/67319/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-08-1100+1b09ca/67319/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-08-1100+1b09ca/67319/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920277
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-08-1100+1b09ca/67319/triggerResults): no differences found

Dr15Jones committed:
Handle case were PuttableSourceBase added to the ProductRegistry

The LHESource doesn't do the standard produces call and instead
directly injects into the ProductRegistry.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44869/summary.html
**COMMIT**: aac9a22ccbd33bdfb7178571356cc92af9e595f3
**CMSSW**: CMSSW_15_1_X_2025-03-09-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47533/44869/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIOPoolInputNoParentDictionary had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-09-0000+1b09ca/67322/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-09-0000+1b09ca/67322/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-09-0000+1b09ca/67322/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920274
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-09-0000+1b09ca/67322/triggerResults): no differences found

Dr15Jones commented:
ignore tests-rejected with manual-override

The failing unit tests is a weakness of the test and will pass when this PR is in the full release.

makortel commented:
+core

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1b09ca/44890/summary.html
**COMMIT**: e2915baf91d6f41e478f54ea9618773f63117a55
**CMSSW**: CMSSW_15_1_X_2025-03-10-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47533/44890/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIOPoolInputNoParentDictionary had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 5 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-10-1100+1b09ca/67345/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 48](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-10-1100+1b09ca/67345/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920232
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-03-10-1100+1b09ca/67345/triggerResults): no differences found

Dr15Jones commented:
ignore tests-rejected with manual-override

civanch commented:
+1

Dr15Jones commented:
ping @cms-sw/dqm-l2 ping

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071

antoniovagnerini commented:
+dqm

Dr15Jones commented:
@cms-sw/orp-l2 please merge as this is needed for work I'm presently doing.

antoniovilela commented:
+1
